lib/Monitor/Ultimagen/RunFolder.pm

```diff
@@ -0,0 +1,327 @@
+      is_paired            => 0,
+      batch_id             => $self->batch_id,
+    };
+
+    my $transaction = sub {
+      $run_row = $rs->create($data);
+      $self->info('Created run ' . $run_row->folder_name . ' with ID ' . $run_row->id_run);
+
+      my $runlane_rs = $run_row->result_source()->schema()->resultset('RunLane');
+      $runlane_rs->create({id_run => $run_row->id_run, position => 1});
+      $self->info("Created record for lane 1 of run_id " . $run_row->id_run);
+    };
+    $rs->result_source()->schema()->txn_do($transaction);
+  }
+  return $run_row;
+}
+
+=head2 tracking_instrument
+
+Record representation of an instrument in the tracking database.
+
+The instrument record is retrieved uniquely (by DB definition).
+If no instrument is found, it exits with error.
+
+Returns a Result::Instrument instance from which instrument properties
+can be retrieved in the DB.
+
+=cut
+has q{tracking_instrument} => (
+  isa           => q{npg_tracking::Schema::Result::Instrument},
+  is            => q{ro},
+  lazy          => 1,
+  builder       => q{_build_tracking_instrument},
+);
+sub _build_tracking_instrument {
+  my $self = shift;
+  my $rs = $self->npg_tracking_schema->resultset('Instrument');
+  my $params = {
+    external_name => $INSTRUMENT_NAME
+  };
+  my @instrument_rows = $rs->search($params)->all();
+
+  my $instrument_count = scalar @instrument_rows;
+  if ($instrument_count == 0) {
+    $self->logcroak('No current instrument found in NPG tracking DB with name '
+      . $INSTRUMENT_NAME);
+  }
+
+  my $instrument_row = $instrument_rows[0];
+  $self->debug('Found instrument ' . $instrument_row->name());  
+  return $instrument_row;
+}
+
+sub _set_tags {
+  my ($self) = shift;
+  my @tags = (
+    'staging',
+    'multiplex'
+  );
+
+  foreach my $tag ( @tags ) {
+    $self->tracking_run()->set_tag($USERNAME, $tag);
+    $self->info("$tag tag is set");
+  }
+}
+
+=head2 is_completed
+
+When the run has been successfully completed
+(UploadCompleted.json exists) return 1, otherwise 0.
+
+=cut
+
+sub is_completed {
+  my $self = shift;
+  if (-f catfile($self->runfolder_path, $RUN_UPLOADED_FILE)) {
+    return 1;
+  }
+  return 0;
+}
+
+=head2 process_run
+
+Core function of the class that is called on the run folder
+periodically to update dynamic properties of a run in the DB.
+If a run record does not exist, it is created.
+The tags and the following statuses are
+checked/assigned accordingly:
+- 'run in progress'   run basecalling is in progress
+- 'run mirrored'      run completed successfully
+Each of the above events saves a time stamp in the DB.
+
+When the run status is one of the following, the function
+will return early:
+- 'run mirrored'
+
+=cut
+sub process_run {
+  my $self = shift;
+  my $run_row = $self->tracking_run();
+  my $current_run_status_obj = $run_row->current_run_status;
+  my $run_uploaded_path = catfile($self->runfolder_path, $RUN_UPLOADED_FILE);
+  
+  if ( ! $current_run_status_obj ) {
```

Comment:
The code in this clause can be moved to where a new run is created into the same transaction.